src/Microsoft.AspNet.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -0,0 +1,206 @@
+                            attribute.Value = resolvedUrl;
+                        }
+                    }
+                    else if (attribute.Value is HtmlString)
+                    {
+                        var htmlStringValue = ((HtmlString)attribute.Value).ToString();
+                        if (TryResolveUrl(
+                                htmlStringValue,
+                                tryEncodeApplicationPath: true,
+                                resolvedUrl: out resolvedUrl))
+                        {
+                            attribute.Value = new HtmlString(resolvedUrl);
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool TryResolveUrl(string url, bool tryEncodeApplicationPath, out string resolvedUrl)
+        {
+            resolvedUrl = null;
+
+            if (url == null)
+            {
+                return false;
+            }
+
+            var urlLength = url.Length;
+
+            // Find the start of the potential application relative URL value.
+            var prefixEndIndex = -1;
+            while (++prefixEndIndex < urlLength && char.IsWhiteSpace(url[prefixEndIndex])) ;
+
+            // Before doing more work, ensure that the URL we're looking at is app relative.
+            if (prefixEndIndex < urlLength - 2 && url[prefixEndIndex] == '~' && url[prefixEndIndex + 1] == '/')
+            {
+                var valueEndIndex = prefixEndIndex - 1;
+                while (++valueEndIndex < urlLength && !char.IsWhiteSpace(url[valueEndIndex])) ;
+
+                var prefix = url.Substring(0, prefixEndIndex);
+                var urlValue = url.Substring(prefixEndIndex, valueEndIndex - prefixEndIndex);
+                var appRelativeUrl = UrlHelper.Content(urlValue);
+                var suffix = url.Substring(valueEndIndex);
+
+                if (tryEncodeApplicationPath)
+                {
+                    var postTildaSlashUrlValue = urlValue.Substring(2);
+
+                    if (!appRelativeUrl.EndsWith(postTildaSlashUrlValue, StringComparison.Ordinal))
+                    {
+                        throw new InvalidOperationException(
+                            Resources.FormatCouldNotResolveApplicationRelativeUrl_TagHelper(
+                                url,
+                                nameof(IUrlHelper),
+                                nameof(IUrlHelper.Content),
+                                typeof(UrlResolutionTagHelper).FullName,
+                                typeof(UrlResolutionTagHelper).GetTypeInfo().Assembly.GetName().Name));
+                    }
+
+                    var applicationPath = appRelativeUrl.Substring(0, appRelativeUrl.Length - postTildaSlashUrlValue.Length);
+                    var encodedApplicationPath = HtmlEncoder.HtmlEncode(applicationPath);
+
+                    resolvedUrl = string.Join(string.Empty, prefix, encodedApplicationPath, postTildaSlashUrlValue, suffix);
```

Comment:
Certainly *logically* `string.Join` and `string.Concat` should be the fastest - they have full information ahead of time. Whereas string interpolation and `string.Format` have only partial information (they have to either parser stuff - ewwww - or otherwise predict the consequences of format specifiers).

But anyway, regarding preserving leading or trailing whitespace: meh, not sure that that's important at all. It's super rare to ever see that, and so what if we don't preserve it.